Add IPFS/Filecoin tutorial using NFT.Storage (#73)
harmony-one/docs-home: developers/tutorials/ipfs-filecoin/using-nftstorage.md

---
description: This tutorial will show how to use NFT.Storage to store NFTs on IPFS and Filecoin for harmony blockchain
---

# Using NFT.Storage

> **NFT.Storage** is a long-term storage service designed for **off-chain** NFT data (like metadata, images, and other assets). Data is [content addressed](https://nftschool.dev/concepts/content-addressing) using **IPFS**, meaning the URL pointing to a piece of data (“ipfs://…”) is completely unique to that data. IPFS URLs can be used in NFTs and metadata to ensure the NFT forever actually refers to the intended data (eliminating things like rug pulls).

Reviewer: Do we want to include a line like: 
IPFS URLs are storage layer agnostic - meaning you can store your data on [many storage layers at once](https://nft.storage/blog/post/2021-12-14-storage-layer-maximalism/) (whether it be on local storage, cloud storage, or multiple decentralized storage networks.

NFT.Storage stores many copies of uploaded data on the public **IPFS** network and **Filecoin** network in two primary ways: in dedicated IPFS servers managed by NFT.Storage, and decentralized on [Filecoin](https://filecoin.io/). Since IPFS is a standard used by many different storage services, it's easy to redundantly store data uploaded to NFT.Storage on any other IPFS-compatible storage solutions.

This tutorial will walk you through the process of creating a simple NFT demo on Harmony, using NFT.Storage to store off-chain NFT data on IPFS and Filecoin to achieve the total decentralization of your NFT.

### Overview

What we will do in this tutorial:

1. Using Truffle to develop a HRC721/NFT smart contract

> 2. Connecting to Harmony blockchain via Metamaks

Reviewer 2: typo Metamaks

   First, install `@truffle/hdwallet-provider` via npm first.

   ```shell
   npm install --save-dev @truffle/hdwallet-provider
   ```

   Add harmony testnet configuration to `truffle-config.js` file. Also make sure to request some ONE token in the wallet so we can pay for transaction fees for smart contract deployment.

   ```javascript
   const HDWalletProvider = require('@truffle/hdwallet-provider');
   const privateKeyTest = '<ADD-YOUR-PRIVATE-KEY-HERE>';//If you have harmony wallet in MetaMask, you can export the private key.
   
   ....<other code>....
   networks: {
       harmonyTestnet: {
         provider: () => {
           if (!privateKeyTest.trim()) {
             throw new Error(
               'Please enter a private key with funds, you can use the default one'
             );
           }
           return new HDWalletProvider({
             privateKeys: [privateKeyTest],
             providerOrUrl: 'https://api.s0.b.hmny.io',
           });
         },
         network_id: 1666700000,
       },
     },
   ... <other code>...
   ```

>    Now the truffle configuration is all set, wen can run the following command to deploy the NFT contract to Harmony Testnet.

Reviewer 2: typo wen

   ```shell
   truffle deploy --network harmonyTestnet
   ```

>    Once the smart contract is deployed success, remember to record your **smart contract address** which we will also use it later. You can also find your NFT contract on [Harmony Testnet blockchain explorer](https://explorer.pops.one/hrc721), also test it on [Remix](https://remix.ethereum.org/).

Reviewer 2: "also use it later" -- remove "it"

### Connect to Metamask 

> In this step, we will write some front-end code to connect to MetaMask wallet so that you can use is to mint NFT and pay transaction fees in next step. 

Reviewer 2: "use is to mint" -> "use it to mint"

Before we move forward, make sure we have installed MetaMask browser extension and add Harmony Network RPC Endpoint. Check more instructions [here](https://docs.harmony.one/home/network/wallets/browser-extensions-wallets/metamask-wallet#1.-adding-a-custom-rpc-endpoint).

Now, let's start coding. 

1. **Create a WalletConnect component**

   We have already created the react app from the first step, so we can directly adding our code into `harmony-nft/src` folder. 

>    Create a new folder call components under `src` folder, then create a `WalletConnect.js` file which we will add our code to detect wallet provider and connect to MetaMask.   [The full WalletConnect.js file can be found on Github](https://github.com/longfeiWan9/harmony-nft-demo/blob/master/src/components/WalletConnect.js).

Reviewer 3: ```suggestion
   Within `harmony-nft/src`, create a new folder `components`. Inside `components`, create a `WalletConnect.js` file and paste in the following code: 
```

>    Also make sure requesting some ONE token in the wallet so we can pay for transaction fees for minting NFT.

Reviewer 2: clarify please

Reviewer 3: harmony's token is $ONE... makes for confusing instructions. i made a suggestion to clarify.

Author: Thank you Mosh for explaining.

>    Also make sure requesting some ONE token in the wallet so we can pay for transaction fees for minting NFT.

Reviewer 3: ```suggestion
   Also make sure to request some $ONE token from the [Harmony One testnet faucet](https://faucet.pops.one/). You'll need some funds to pay for transaction fees for minting the NFT.
```

   //The core code to detect wallet provider and connect to MetaMask
   const connectWalletHandler = async() => {
   		if (window.ethereum && window.ethereum.isMetaMask) {
   			console.log('MetaMask Here!');
   
   			try {
   				const accounts = await window.ethereum.request({ method: 'eth_requestAccounts'});
   				accountChangedHandler(accounts[0]);
   				setConnButtonText('Wallet Connected');
   				setWalletConnected(true);
   			} catch (error) {
   				setErrorMessage(error.message);
   			}
   
   		} else {
               setWalletConnected(false);
   			console.log('Need to install MetaMask');
   			setErrorMessage('Please install MetaMask browser extension to interact');
   		}
   	}
   ```

2. **Add the WalletConnect component to `App.js`**

   ```javascript
   import WalletConnect from './components/WalletConnect'
   ...<other code>...
   function App() {
     return (
       <div className='App'>
         <div className='logos'>
           <img className='harmony' src={harmonyLogo} alt="Harmony Logo"/>
           <img className='nftStorage' src={nftStorageLogo} alt="Harmony Logo"/>
         </div>
         <WalletConnect/>
       </div>
     );
   }
   ```

>    Now we have the code ready, then let's test the Wallet connection function. First run the following command to start the react application.

Reviewer 3: ```suggestion
   Now, let's test the wallet connection function. Run the following command to start the React application.
```

   ```shell
   npm start
   ```

>    Once the application is up, you can click the `Connect Wallet`button to invoke the MetaMask and authorize the connection.

Reviewer 2: missing space before "button"

>    Once the application is up, you can click the `Connect Wallet`button to invoke the MetaMask and authorize the connection.

Reviewer 3: ```suggestion
   This should open your browser and load `localhost:3000`. Once the application is running, you can click the `Connect Wallet` button to invoke MetaMask and authorize the connection.
```

   ![Connect to MetaMask](https://bafybeidngoqv4qq46bcb434em6skbg2b4ogyate3fwewutro67pm2j34qy.ipfs.dweb.link/HarmonyNFT-WalletConnect.png)

   

### MintNFT process

> Before this step, we should have the following items ready to write code to mint NFT.

Reviewer 3: ```suggestion
In this step, we will write the code that mints the NFT. Have the following items ready:
```

> + NFT Smart Contract deployed to Harmony TestNet. 

Reviewer 3: ```suggestion
+ NFT Smart Contract deployed to Harmony Testnet. 
```

  + NFT contract address
  + NFT contract manifest JSON file

> + Metamask browser extension installed and filled your wallet with some ONE token.

Reviewer 3: ```suggestion
+ Metamask browser extension installed, and a wallet with at least 1 Harmony ONE testnet token.
```

> In this step, we will write code to mint a NFT on Harmony blockchain and also upload your NFT asset and metadata to IPFS & Filecoin so that your NFT is actually truly decentralized other than storing on centralized cloud. Here, we will use a tool called [NFT.Storage](https://nft.storage/) which offers free and long-term storage on **IPFS** and **Filecoin**.

Reviewer 2: remove "actually"

   ![NFT Storage API Key](https://bafybeibckdbu5uorpl6itmukw7o35zq3xsvzeynwxdkbl3azrd64xwtgfm.ipfs.dweb.link/NFTStorage-API-KEYS.png)

   Also we will use ethers to interact with MetaMask to invoke NFT smart contract which is deployed on Harmony blockchain. 

2. **Create mintNFT component**

   Now we  will create a `MintNFT.js` component in `src/components` folder. Let's first create a form to upload NFT Images in `MintNFT.js`.

   ```javascript
   import React, {useState} from 'react';
   const MintNFT =() => {
     const [uploadedFile, setUploadedFile] = useState();
   
     const handleFileUpload = (event) => {
         setUploadedFile(event.target.files[0]);
     }
   
     const mintNFTToken = async(inputFile) =>{
         //Will add MintNFT logic here.
     }
   
     return(
         <div className='MintNFT'>
             <form>
                 <h3>Mint your NFT on Harmony & Filecoin/IPFS</h3>
                 <input type="file" onChange={handleFileUpload}></input>
                 <button onClick={e=>mintNFTToken(e, uploadedFile)}>Mint NFT</button>
             </form>
         </div>
     )
   }
   export default MintNFT;
   ```
   

>    Once the NFT image is captured via form upload, we can upload NFT asset and metadata to IPFS/Filecoin via NFT.storage, then invoke NFT smart contract on Harmony to mint a new item. The basic logic will be like following:

Reviewer 3: ```suggestion
   Once the NFT image is captured via form upload, we will upload the NFT asset and metadata to IPFS/Filecoin via NFT.storage, then invoke the NFT smart contract on Harmony to mint a new item. The basic logic will be like following:
```

   
   ```javascript
   const mintNFTToken = async(event, uploadedFile) =>{

>        //1. upload NFT content via NFT.storage

Reviewer 3: ```suggestion
       //1. Upload NFT content via NFT.storage
```

       const metaData = await uploadNFTContent(uploadedFile);
   
       //2. Mint a NFT token on Harmony
       const mintNFTTx = await sendTxToHarmony(metaData);
   }
   ```
   
3. **Upload NFT asset via NFT.Storage**

>    Now let's set up NFT.Storage to upload NFT assets.

Reviewer 3: ```suggestion
   Now let's set up NFT.Storage so you can upload NFT assets.
```

>    + Install NFT.Storage dependency in your react app.

Reviewer 3: ```suggestion
   + Install NFT.Storage dependency in your React app.
```

     ```shell
     npm install nft.storage
     ```

>    + We are all set to start coding your logic, we need to first import NFTstorage and get the API Key that we created in the first step.

Reviewer 3: ```suggestion
   + We are all set to start coding your logic. Add the following code blocks to `MintNFT.js`. (You can also [copy the entire sample file from Github](https://github.com/longfeiWan9/harmony-nft-demo/blob/master/src/components/MintNFT.js).)
   
    First, import NFTstorage and get the API Key that we created in the first step.
```

     ```javascript

>      //Add this in MintNFT.js

Reviewer 3: ```suggestion
```

     import { NFTStorage } from "nft.storage";
     const APIKEY = '<Your-API-KEY>';
     ...
     ```

>    + Then initializing NFTStorage connection using your API KEY.

Reviewer 3: ```suggestion
   + Then, initialize the NFTStorage connection using your API KEY.
```

     ```javascript
     const uploadNFTContent = async(inputFile) =>{

>      		//Initialize NFTStorage

Reviewer 3: ```suggestion
     	//Initialize NFTStorage
```

         const nftStorage = new NFTStorage({token: APIKEY,});
     }
     ```

>    + We can simply call `nftStorage.store()` to upload NFT content, which will upload your NFT as well as its related metadata to IPFS & Filecoin. You can customized your NFT metadata by updating properties like name and description.

Reviewer 2: "customized" => "customize"

>    + We can simply call `nftStorage.store()` to upload NFT content, which will upload your NFT as well as its related metadata to IPFS & Filecoin. You can customized your NFT metadata by updating properties like name and description.

Reviewer 3: ```suggestion
   + To upload NFT content, we'll call `nftStorage.store()`, which will upload your NFT as well as its related metadata to multiple, redundant IPFS & Filecoin nodes. You can customize your NFT metadata by updating properties like name and description.
```

     ```javascript
     const uploadNFTContent = async(inputFile) =>{
         //Initialize NFTStorage
         const nftStorage = new NFTStorage({token: APIKEY,});
         try {
             //Upload NFT to IPFS & Filecoin
             const metadata = await nftStorage.store({
                 name: 'Harmony NFT collection',
                 description: 'This is a Harmony NFT collenction stored on IPFS & Filecoin.',
                 image: file
             });
             return metadata;
     
         } catch (error) {
             setErrorMessage("Could not save NFT to NFT.Storage - Aborted minting.");
             console.log(error);
         }
     }
     ```
   
     If NFT asset is uploaded successfully, NFT.Storage will do the following two things:

     + Store NFT image on IPFS & Filecoin.

>      + Update NFT metadata with IPFS Link to your NFT image, then store NFT metadata on IPFS & Filecoin.

Reviewer 3: ```suggestion
     + Update NFT metadata with the IPFS link to your NFT image, then also store the NFT metadata on IPFS & Filecoin.
```

   
4. **Mint a new NFT on Harmony**

>    After getting the Metadata on IPFS, we can invoke the NFT smart contract on Harmony to mint a new NFT using NFT metadata on IPFS as tokenURI.

Reviewer 3: ```suggestion
   After getting the metadata on IPFS, we can invoke the NFT smart contract on Harmony to mint a new NFT using that metadata as tokenURI.
```

   In this step, we will need to use `ethers` to interact with Metamask, so we first install this dependency.

   ```shell
   npm install ethers
   ```

   In order to invoke the NFT contract we have deployed to Harmony Testnet, we need to grab the contract ABI from the manifest JSON file which was generated via smart contract compilation.

   ```shell
   //run this command under harmony-nft folder
   cp ../build/contracts/HarmonyNFT.json src/abi/
   ```

>    Then we are ready to write code to send MintNFT transaction in `MintNFT.js` . First import ethers and smart contract manifest file,  and aslo create a `nftContractAddress` for the deployed NFT contract. 

Reviewer 3: ```suggestion
   Then we are ready to write code to send MintNFT transaction in `MintNFT.js` . This will import ethers and the smart contract manifest file, and also create a `nftContractAddress` for the deployed NFT contract. 
```

   ```javascript
   import HarmonyNFT from "../abi/HarmonyNFT.json";
   import { ethers } from 'ethers';
   
   const nftContractAddress = '0x494543afa8445025B9dBc7a39EdC7Fc624cA46F6';
   ```

   Then we can load the deployed NFT contract and invoke mintItem method.

   ```javascript
   //Add this code to load deployed NFT contract and invoke mintItem method.
   const sendTxToHarmony = async(metadata) =>{
       try {
           const provider = new ethers.providers.Web3Provider(window.ethereum);
           const connectedContract = new ethers.Contract(
               nftContractAddress,
               HarmonyNFT.abi,
               provider.getSigner()
           );
           const mintNFTTx = await connectedContract.mintItem(metadata.url);
           console.log(mintNFTTx);
       } catch (error) {
           setErrorMessage("Failed to send tx to Harmony.");
           console.log(error);
       }
   }
   ```

5. **Preview the new minted NFT**

>    By now, we have all the codes to upload NFT data, send mint item transaction to Harmony. Let's add some code to preview our fresh minted NFT, including the NFT image, NFT metadata on IPFS and mintNFT transaction on Harmony.

Reviewer 3: ```suggestion
   By now, we have all the code to upload NFT data and send the mint item transaction to Harmony. Let's add some code to preview our freshly minted NFT, including the NFT image, NFT metadata on IPFS, and mintNFT transaction on Harmony.
```

>    You can find the full [MintNFT.js file on Github](https://github.com/longfeiWan9/harmony-nft-demo/blob/master/src/components/MintNFT.js).

Reviewer 3: This transition feels a little bit unexpected. Are we telling people to open this file again?

Author: It is actually some extra code added in `MintNFT.js`, so I will  recommend developers to check the full file to see if they are writing the code correctly.

         setImageView(imgViewString);
         setMetaDataURl(getIPFSGatewayURL(metaData.url));
         setTxURL('https://explorer.pops.one/tx/'+ mintNFTTx.hash);
         setTxStatus("NFT is minted successfully!");
     }
   
     //
     const getIPFSGatewayURL = (ipfsURL)=>{
         let urlArray = ipfsURL.split("/");
         let ipfsGateWayURL = `https://${urlArray[2]}.ipfs.dweb.link/${urlArray[3]}`;
         return ipfsGateWayURL;
     }
     ...
     return(
         <div className='MintNFT'>
             <form>
                 <h3>Mint your NFT on Harmony & Filecoin/IPFS</h3>
                 <input type="file" onChange={handleFileUpload}></input>
                 <button onClick={e=>mintNFTToken(e)}>Mint NFT</button>
             </form>
             {errorMessage}
             {txStatus && <p>{txStatus}</p>}
             {imageView && <img className='NFTImg' src={imageView} alt="NFT image preview"/>}
             {metaDataURL && <p><a href={metaDataURL}>Metadata on IPFS</a></p>}
             {txURL && <p><a href={txURL}>See the mint transaction</a></p>}
         </div>
     )
   }
   ```

### Let's run it

> Now we have everything finished, we can run `npm start` to start our react application, then we will have this page if everything works fine. 

Reviewer 3: ```suggestion
Now that we've finished writing all the code, we can run `npm start` to start our React application. If everything is working, you should see a page like this:
```

> ![image-20220128113617250](/Users/longfeiwang/Library/Application Support/typora-user-images/image-20220128113617250.png)

Reviewer 2: this one is broken
